parquet/src/test/java/org/apache/iceberg/parquet/TestInternalWriter.java

```diff
@@ -0,0 +1,132 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.parquet;
+
+import static org.apache.iceberg.parquet.ParquetWritingTestUtils.createTempFile;
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.UUID;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileContent;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Files;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.data.parquet.InternalReader;
+import org.apache.iceberg.data.parquet.InternalWriter;
+import org.apache.iceberg.expressions.Literal;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructProjection;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+public class TestInternalWriter {
+  @TempDir private Path temp;
+
+  @Test
+  public void testDataWriter() throws IOException {
+    Schema schema =
+        new Schema(
+            required(100, "b", Types.BooleanType.get()),
+            optional(101, "i", Types.IntegerType.get()),
+            required(102, "l", Types.LongType.get()),
+            optional(103, "f", Types.FloatType.get()),
+            required(104, "d", Types.DoubleType.get()),
+            optional(105, "date", Types.DateType.get()),
+            required(106, "time", Types.TimeType.get()),
+            required(107, "ts", Types.TimestampType.withoutZone()),
+            required(108, "ts_tz", Types.TimestampType.withZone()),
+            required(109, "s", Types.StringType.get()),
+            required(110, "uuid", Types.UUIDType.get()),
+            required(111, "fixed", Types.FixedType.ofLength(7)),
+            optional(112, "bytes", Types.BinaryType.get()),
+            required(113, "dec_38_10", Types.DecimalType.of(38, 10)));
+
+    // Consuming the data as per Type.java
+    GenericRecord record = GenericRecord.create(schema);
```

Comment:
Should I introduce a new class like `GenericStruct` that implements `StructLike`, I thought of using `StructProjection` but it is read only. Hence, need one simple class to create a `StructLike` objects. 

Line 95, wrapping this with `StructProjection` just to get pure `StructLike`